File: bezier/src/org/brogan/bezier/OvalManager.java
package org.brogan.bezier;

import java.awt.*;
import java.awt.geom.*;

/**
 * Represents a single axis-aligned ellipse on the canvas.
 * Coordinates are in the BezierDrawPanel pixel space (0..GRIDWIDTH/GRIDHEIGHT
 * with 20 px edgeOffset border). Normalisation to -0.5..+0.5 happens at save time.
 *
 * Ovals participate in Polygon Selection mode (select/move/scale/rotate/flip)
 * but do NOT support point or edge sub-selection.
 */
public class OvalManager {

    private double cx;   // centre x in canvas pixel space
    private double cy;   // centre y in canvas pixel space
    private double rx;   // x-radius in canvas pixel space
    private double ry;   // y-radius in canvas pixel space
    // Frozen values for scale gestures — set once at gesture start, like CubicPoint.origPos
    private double origCx;
    private double origCy;
    private double origRx;
    private double origRy;
    private int layerId = 0;
    private boolean selected = false;

    public OvalManager(double cx, double cy, double rx, double ry) {
        this.cx = cx;
        this.cy = cy;
        this.rx = rx;
        this.ry = ry;
        this.origCx = cx;
        this.origCy = cy;
        this.origRx = rx;
        this.origRy = ry;
    }

    // ── Accessors ──────────────────────────────────────────────────────────

    public double getCx() { return cx; }
    public double getCy() { return cy; }
    public double getRx() { return rx; }
    public double getRy() { return ry; }
    public double getOrigCx() { return origCx; }
    public double getOrigCy() { return origCy; }
    public void setCx(double v) { cx = v; }
    public void setCy(double v) { cy = v; }
    public void setRx(double v) { rx = v; }
    public void setRy(double v) { ry = v; }
    public int  getLayerId()    { return layerId; }
    public void setLayerId(int id) { layerId = id; }
    public boolean isSelected() { return selected; }
    public void setSelected(boolean s) { selected = s; }

    // ── Geometry ──────────────────────────────────────────────────────────

    /** True if the given canvas-space point is inside this ellipse. */
    public boolean contains(double px, double py) {
        double dx = (px - cx) / Math.max(rx, 1.0);
        double dy = (py - cy) / Math.max(ry, 1.0);
        return dx * dx + dy * dy <= 1.0;
    }

    /** Translate centre by (dx, dy). */
    public void translate(double dx, double dy) {
        cx += dx;
        cy += dy;
    }

    /**
     * Freeze current values as the origin for this scale gesture.
     * Must be called once at the start of each slider gesture (equivalent to
     * CubicPoint.setOrigPos used by polygon scaling).
     */
    public void freezeOrig() {
        origCx = cx; origCy = cy; origRx = rx; origRy = ry;
    }

    /**
     * Scale around the given pivot using the FROZEN origin values.
     * Safe to call repeatedly with different factors — always relative to the
     * state captured by freezeOrig(), so slider direction reversal works correctly.
     */
    public void scaleXYFromOrig(double factor, double pivotX, double pivotY) {
        cx = pivotX + (origCx - pivotX) * factor;
        cy = pivotY + (origCy - pivotY) * factor;
        rx = Math.abs(origRx * factor);
        ry = Math.abs(origRy * factor);
    }

    public void scaleXFromOrig(double factor, double pivotX) {
        cx = pivotX + (origCx - pivotX) * factor;
        rx = Math.abs(origRx * factor);
    }

    public void scaleYFromOrig(double factor, double pivotY) {
        cy = pivotY + (origCy - pivotY) * factor;
        ry = Math.abs(origRy * factor);
    }

    /**
     * Scale around the given pivot point (cumulative — avoid in slider gestures).
     * Scales both centre position and radii uniformly (XY) or on one axis.
     */
    public void scaleXY(double factor, double pivotX, double pivotY) {
        cx = pivotX + (cx - pivotX) * factor;
        cy = pivotY + (cy - pivotY) * factor;
        rx = Math.abs(rx * factor);
        ry = Math.abs(ry * factor);
    }

    public void scaleX(double factor, double pivotX) {
        cx = pivotX + (cx - pivotX) * factor;
        rx = Math.abs(rx * factor);
    }

    public void scaleY(double factor, double pivotY) {
        cy = pivotY + (cy - pivotY) * factor;
        ry = Math.abs(ry * factor);
    }

    /** Rotate centre around the given pivot by degrees. Radii unchanged (oval stays axis-aligned). */
    public void rotate(double degrees, double pivotX, double pivotY) {
        double rad = Math.toRadians(degrees);
        double dx = cx - pivotX;
        double dy = cy - pivotY;
        double cosA = Math.cos(rad);
        double sinA = Math.sin(rad);
        cx = pivotX + dx * cosA - dy * sinA;
        cy = pivotY + dx * sinA + dy * cosA;
    }

    /** Flip horizontally around the given x-centre. */
    public void flipH(double centerX) {
        cx = 2.0 * centerX - cx;
    }

    /** Flip vertically around the given y-centre. */
    public void flipV(double centerY) {
        cy = 2.0 * centerY - cy;
    }

    // ── Rendering ─────────────────────────────────────────────────────────

    /**
     * Draw the oval into the given Graphics2D.
     * Selected ovals are drawn with a 3 px coloured outline.
     * Unselected ovals are drawn with the given stroke colour.
     */
    public void draw(Graphics2D g2D, Color strokeColor,
                     boolean discrete, boolean relational) {
        double x = cx - rx;
        double y = cy - ry;
        double w = rx * 2;
        double h = ry * 2;

        RenderingHints rh = new RenderingHints(
            RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
        g2D.setRenderingHints(rh);

        if (selected) {
            Color selColor = relational
                ? new Color(255, 140, 0)   // orange — relational
                : new Color(100, 150, 255); // blue — discrete
            g2D.setColor(selColor);
            g2D.setStroke(new BasicStroke(4f));
            g2D.draw(new Ellipse2D.Double(x, y, w, h));
        } else {
            g2D.setColor(strokeColor);
            g2D.setStroke(new BasicStroke(1.5f));
            g2D.draw(new Ellipse2D.Double(x, y, w, h));
        }
    }

    /** Deep copy for undo / clipboard use. */
    public OvalManager copy() {
        OvalManager m = new OvalManager(cx, cy, rx, ry);
        m.layerId = this.layerId;
        m.selected = false;
        return m;
    }
}

File: bezier_py/model/oval_manager.py
"""
OvalManager — one axis-aligned ellipse in canvas pixel space.
Mirrors Java OvalManager.
Coordinates: canvas pixel space (0..1040), centre at (520,520).
"""
from __future__ import annotations
import math
import copy


class OvalManager:
    """
    cx, cy : centre in canvas pixel space
    rx, ry : radii in canvas pixel space
    orig*  : frozen values for slider-based scale gestures (like CubicPoint.orig_pos)
    """

    def __init__(self, cx: float, cy: float, rx: float, ry: float) -> None:
        self.cx = cx
        self.cy = cy
        self.rx = rx
        self.ry = ry
        self.orig_cx = cx
        self.orig_cy = cy
        self.orig_rx = rx
        self.orig_ry = ry
        self.layer_id: int = 0
        self.selected: bool = False

    # ── geometry ──────────────────────────────────────────────────────────────

    def contains(self, px: float, py: float) -> bool:
        """True if canvas-space point is inside this ellipse."""
        rx = max(self.rx, 1.0)
        ry = max(self.ry, 1.0)
        dx = (px - self.cx) / rx
        dy = (py - self.cy) / ry
        return dx * dx + dy * dy <= 1.0

    def translate(self, dx: float, dy: float) -> None:
        self.cx += dx
        self.cy += dy

    def freeze_orig(self) -> None:
        """Capture current values at gesture start (for slider-based scale)."""
        self.orig_cx = self.cx
        self.orig_cy = self.cy
        self.orig_rx = self.rx
        self.orig_ry = self.ry

    def scale_xy_from_orig(self, factor: float, pivot_x: float, pivot_y: float) -> None:
        self.cx = pivot_x + (self.orig_cx - pivot_x) * factor
        self.cy = pivot_y + (self.orig_cy - pivot_y) * factor
        self.rx = abs(self.orig_rx * factor)
        self.ry = abs(self.orig_ry * factor)

    def rotate(self, degrees: float, pivot_x: float, pivot_y: float) -> None:
        """Rotate centre around pivot. Radii unchanged (stays axis-aligned)."""
        rad = math.radians(degrees)
        dx = self.cx - pivot_x
        dy = self.cy - pivot_y
        self.cx = pivot_x + dx * math.cos(rad) - dy * math.sin(rad)
        self.cy = pivot_y + dx * math.sin(rad) + dy * math.cos(rad)

    def flip_h(self, center_x: float) -> None:
        self.cx = 2.0 * center_x - self.cx

    def flip_v(self, center_y: float) -> None:
        self.cy = 2.0 * center_y - self.cy

    def average_xy(self):
        from PySide6.QtCore import QPointF
        return QPointF(self.cx, self.cy)

    def copy(self) -> OvalManager:
        m = OvalManager(self.cx, self.cy, self.rx, self.ry)
        m.layer_id = self.layer_id
        m.selected = False
        return m
